### DISCUSSION TITLE
ISO filename, checksum and hash method

### DISCUSSION THREAD

**Discussion by sestegra** [vmware/packer-examples-for-vsphere#72] [Ideas]
Hello,

I like how the project split global and specific distribution configurations in Packer files.
Nevertheless, I suggest to add the ISO configuration level (use-cases and explanations below).

The configuration levels would be
-  **Global configuration** stored in `config` folder (with defaults copied from `builds/*.pkrvars.hcl`)
- **Specific distribution configuration** located in the related distribution folder in `builds/<OS family>/<OS version>`
- **ISO configuration** located in a place to be defined (that changes often)

So, my concern is about ISO filename, checksum and hash method.

First of all, I prefer to use the exact filename (chosen by the distribution vendor) to save it on the datastore.
In fact, a Google search of that filename allows quicker search on it, when required.
Then, I would like to keep whole releases on the datastore (e.g. `ubuntu-20.04.2-live-server-amd64.iso`, `ubuntu-20.04.3-live-server-amd64.iso`).

Secondly, I prefer to use the checksum values that are GPG verified by the distribution vendor.
By this way, the hash method couldn't be global, but specific: no whole vendor use the same hash method.

Thirdly, I suggest to remove the ISO checksum from the specific distribution configuration.
In fact, I see 2 use-cases:
- the full version or the minimal version of the ISO could be used to build an image
- multiple release of the same OS version could be used to build an image

So, my suggestion is composed by following updates:
- Remove `common_iso_hash` variable, so integrate the hash method in `iso_checksum` variable
- Remove `iso_filename` and `iso_checksum` from the `specific distribution configuration`
- Add `iso_filename` and `iso_checksum` in `build.sh` (or files loaded by `build.sh`)

As result, the **Specific distribution configuration** stays stable without any update when the vendor releases a new version.
**ISO configuration** would become flexible and customizable like **Global configuration**.

What do you think about it?

Kind regards,
Stéphane

**Comment by sestegra** [comment]
For Windows 10 Professional, the installer ISO contains only one language.

So, **ISO configuration** would contain `vm_inst_os_language` and `vm_inst_os_keyboard` to use that language.
For instance, the English International version use `en-GB`.

**Comment by tenthirtyam** [comment]
Some refactoring in #93. Plan to re-org the `builds` directory structure similar to below:


```
├── builds
│   ├── linux
│   │   └── distribution
│           └── version
│   │           ├── *.pkr.hcl
│   │           ├── *.auto.pkrvars.hcl
│   │           └── data
│   │               └── ks.pkrtpl.hcl
│   └── windows
│       └── windows
│           └── version
│               ├── *.pkr.hcl
│               ├── *.auto.pkrvars.hcl
│               └── data
│                   └── autounattend.pkrtpl.hcl
│       └── windows-server
│           └── version
│               ├── *.pkr.hcl
│               ├── *.auto.pkrvars.hcl
│               └── data
│                   └── autounattend.pkrtpl.hcl
```

Let's discuss this week.

Ryan

**Reply by tenthirtyam** [reply]
+ @gcblack for thoughts, too.

**Reply by sestegra** [reply]
Look good to me for the re-org of the `builds` directory.

However (in #93), if `./build.sh` is launched with all, I can't change ISO file and checksum for more than one target.
